File: NumberIsland.java
/*
DFS : Time complexity : O(N*M) matrix size
Space complexity: O(1)


*/

class Solution {
    int dirs [][];
    int m;
    int n;
    char [][] grid;
    public int numIslands(char[][] grid) {
        
        dirs= new int[][] {{1,0},{-1,0},{0,1},{0,-1}};
        this.grid = grid;
        if(grid == null)  return 0;
        m = grid.length;
        n = grid[0].length;
        int count = 0;
        for(int i=0;i < m;i++) {
            for(int j=0; j < n ; j++) {
                if(grid[i][j] == '1') {
                    count++;
                    grid[i][j]  = '0';
                    dfs(i,j);   
                }   
            }     
        }
        
        
        return count;
    }
    
    
    private void dfs(int row, int col) {
        
        for(int [] dir: dirs) {
            int r = dir[0];
            int c = dir[1];    
            int newR = row + r;
            int newC = col + c;
            if(newR >= 0 && newR < m && newC >=0 && newC < n && grid[newR][newC] == '1') {
                grid[newR][newC] = '0';
                dfs(newR, newC);   
            }   
        }   
    }
}

File: NumberIsland.py
#200. Number of Islands
"""
Time Complexity : O(m*n) #going through all elements of matrix
Space COmplexity : O(m*n) #stack
"""
class Solution:
    dirs = [[0,1], [0,-1], [-1,0], [1,0]]
    #         R       L       U      D
    
    def dfs(self, grid, r, c):
        #base
        if r < 0 or r > len(grid)-1 or c < 0 or c > len(grid[0])-1 or grid[r][c] == '0' :
            return
        
        #logic
        grid[r][c] = '0' #mark as visited
        for d in self.dirs:
            nr = r + d[0]
            nc = c + d[1]
            
            self.dfs(grid, nr, nc)
        
        
    def numIslands(self, grid: List[List[str]]) -> int:
        #null case
        if len(grid) == 0:
            return 0
        
        
        count = 0
        
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                
                if grid[r][c] == '1':
                    count += 1
                    self.dfs(grid, r, c)
                    
        return count
                    
        
